src/coreclr/jit/lowerxarch.cpp

```diff
@@ -1447,6 +1454,149 @@ void Lowering::LowerHWIntrinsicCmpOp(GenTreeHWIntrinsic* node, genTreeOps cmpOp)
     LowerNode(node);
 }
 
+//----------------------------------------------------------------------------------------------
+// Lowering::LowerHWIntrinsicCndSel: Lowers a Vector128 or Vector256 Conditional Select call
+//
+//  Arguments:
+//     node - The hardware intrinsic node.
+//
+void Lowering::LowerHWIntrinsicCndSel(GenTreeHWIntrinsic* node)
+{
+    var_types   simdType            = node->gtType;
+    CorInfoType simdBaseJitType     = node->GetSimdBaseJitType();
+    var_types   simdBaseType        = node->GetSimdBaseType();
+    unsigned    simdSize            = node->GetSimdSize();
+    bool        isSimdAsHWIntrinsic = node->IsSimdAsHWIntrinsic();
+
+    assert(varTypeIsSIMD(simdType));
+    assert(varTypeIsArithmetic(simdBaseType));
+    assert(simdSize != 0);
+
+    // Get the three arguments to ConditionalSelect we stored in node
+    // op1: the condition vector
+    // op2: the left vector
+    // op3: the right vector
+    GenTree* op1 = node->Op(1);
+    GenTree* op2 = node->Op(2);
+    GenTree* op3 = node->Op(3);
+
+    // If the condition vector comes from a hardware intrinsic that
+    // returns a per element mask (marked with HW_Flag_ReturnsPerElementMask),
+    // we can optimize the entire conditional select to a single BlendVariable instruction
+
+    // First, determine if the target architecture supports BlendVariable
+    bool           supportsBlendVariable = false;
+    NamedIntrinsic blendVariableId;
+
+    if (simdSize == 32)
+    {
+        // for Vector256, BlendVariable for floats is available on AVX, whereas other types require AVX2
+        if (varTypeIsFloating(simdBaseType))
+        {
+            // This should have already been confirmed
+            assert(comp->compIsaSupportedDebugOnly(InstructionSet_AVX));
+            supportsBlendVariable = true;
+            blendVariableId       = NI_AVX_BlendVariable;
+        }
+        else
+        {
+            supportsBlendVariable = comp->compOpportunisticallyDependsOn(InstructionSet_AVX2);
+            blendVariableId       = NI_AVX2_BlendVariable;
+        }
+    }
+    else
+    {
+        // for Vector128, BlendVariable is available on SSE41
+        supportsBlendVariable = comp->compOpportunisticallyDependsOn(InstructionSet_SSE41);
+        blendVariableId       = NI_SSE41_BlendVariable;
+    }
+
+    if (op1->OperIsHWIntrinsic() && supportsBlendVariable)
```

Comment:
A downside of this ordering is that the method will be marked as "opportunistically using" SSE4.1 (Vector128) even if we could never have emitted `blendVariableId`

It might be better to change this slightly to be:
```cpp
NamedIntrinsic blendVariableId = NI_Illegal;

if (op1->OperIsHWIntrinsic() && HWIntrinsicInfo::ReturnsPerElementMask(op1->AsHWIntrinsic()->GetHWIntrinsicId()))
{
    if (simdSize == 32)
    {
        if (varTypeIsFloating(simdBaseType))
        {
            assert(comp->compIsaSupportedDebugOnly(InstructionSet_AVX));
            blendVariableId = NI_AVX_BlendVariable;
        }
        else if (comp->compOpportunisticallyDependsOn(InstructionSet_AVX2))
        {
            blendVariableId = NI_AVX2_BlendVariable;
        }      
    }
    else if (comp->compOpportunisticallyDependsOn(InstructionSet_SSE41))
    {
        blendVariableId = NI_SSE41_BlendVariable;
    }
}

if (blendVariableId != NI_Illegal)
{
    // result = BlendVariable op3 (right) op2 (left) op1 (mask)
    node->ResetHWIntrinsicId(blendVariableId, comp, op3, op2, op1);
    return;
}
```

or something similar that ensures we only mark it as "opportunistic" when we know we could actually use the instruction